sql/src/main/java/org/apache/druid/sql/avatica/DruidConnection.java

```diff
@@ -53,25 +54,29 @@
   private final AtomicInteger statementCounter = new AtomicInteger();
   private final AtomicReference<Future<?>> timeoutFuture = new AtomicReference<>();
 
-  @GuardedBy("statements")
-  private final Map<Integer, DruidStatement> statements;
+  // Typically synchronized by connectionLock, except in one case: the onClose function passed
+  // into DruidStatements contained by the map.
+  private final ConcurrentMap<Integer, DruidStatement> statements;
 
-  @GuardedBy("statements")
+  @GuardedBy("connectionLock")
```

Comment:
Hi @asdf2014 , the field connectionLock now needs to be accessed synchronized, So I think should add an annotation here

Reply:
To me, it makes sense because `statements` is now _not_ always guarded by `connectionLock`. But, it is in some (well, many) cases. The comment explains it in a way that the annotation would struggle to.

Reply:
Thanks for explaining this @gianm @zhaojiandong . I mean we shouldn't remove the `@GuardedBy("connectionLock")` from `statements` because the `statements` is not always synchronized by `connectionLock` and the CI fails, which is unreasonable. In addition, putting a `@GuardedBy` on `connectionLock` is also not necessary. If this is done, it means that we should have the `connectionLock` lock object when we use `connectionLock`, in fact it cannot achieve the purpose that we should own the `connectionLock` when using `statements`. So, i think there is a better way to solve this situation by adding a `@SuppressWarnings("FieldAccessNotGuarded")` annotation to the `Runnable#run` method, or adding a `// noinspection FieldAccessNotGuarded` comment to the `statements.remove(statementId)` caluse, which will suppress this check and make CI happy.

```java
@GuardedBy("connectionLock")
private final ConcurrentMap<Integer, DruidStatement> statements;
private final Object connectionLock = new Object();

public DruidStatement createStatement(SqlLifecycleFactory sqlLifecycleFactory)
{
    // ...
    final DruidStatement statement = new DruidStatement(
        // ...
        () -> {
          // noinspection FieldAccessNotGuarded
          statements.remove(statementId);
        }
    );
    // ...
}
```

Reply:
Ah, I see what you mean now @asdf2014 -- your patch #6903 looks good, thanks!

Reply:
@gianm You are welcome, I am sorry for my poor English expression :cold_sweat: